### PR TITLE
[The Space] add getTokensByOwner method

### PR DESCRIPTION
**Pull request by gary02**
Note that this impl may not be gas-optimized  because of using ERC721Enumerable.

### CODE REVIEW

File: src/TheSpace/HarbergerMarket.sol [hunk 1]
```diff
@@ -155,7 +164,7 @@ contract HarbergerMarket is ERC721, IHarbergerMarket, Multicall, AccessRoles {
     /**
      * @dev See {IERC20-totalSupply}. Always return total possible amount of supply, instead of current token in circulation.
      */
-    function totalSupply() public view returns (uint256) {
+    function totalSupply() public view override(ERC721Enumerable) returns (uint256) {
         return _totalSupply;
     }
```
Comment (robertu7) [inline src/TheSpace/HarbergerMarket.sol:169]: Add these too?

```solidity
    function supportsInterface(bytes4 interfaceId)
        public
        view
        override(ERC721, ERC721Enumerable)
        returns (bool)
    {
        return super.supportsInterface(interfaceId);
    }
```

File: src/TheSpace/TheSpace.sol [hunk 2]
```diff
@@ -79,10 +79,38 @@ contract TheSpace is HarbergerMarket {
...
     function getColor(uint256 tokenId) public view returns (uint256) {
         return pixelColor[tokenId];
     }
+
+    /**
+     * @notice Get owned tokens for a user.
+     *
+     * @dev offset based pagination
+     */
+    function getTokensByOwner(
```
Comment (robertu7) [inline src/TheSpace/TheSpace.sol:92]: Shall we add another interface to batch query pixels?

Comment (gary02) [reply src/TheSpace/TheSpace.sol:92]: Do you mean adding `getPixelsByOwner` besides adding `getTokensByOwner` ？

File: src/TheSpace/HarbergerMarket.sol (hunk 1 above)
Comment (gary02) [reply src/TheSpace/HarbergerMarket.sol:169]: Yes, I will update this.

Comment (gary02) [reply src/TheSpace/HarbergerMarket.sol:169]: > Yes, I will update these.

Oh, I just find that I have added `override ERC721Enumerable` in  `supportsInterface`, see https://github.com/gary02/contracts/blob/5f1cd251a5d9efb94cbce9796010d4fa9206e966/src/TheSpace/HarbergerMarket.sol#L107

**Comment by gary02** [comment]
I think there should be a custom erc721 contract inheriting `ERC721Enumerable`  and inherited by `HarbergerMarket`. `getTokensByOwner` and some methods/state variables in  `HarbergerMarket` should be moved to this erc721 contract. @guoliu @robertu7

File: src/TheSpace/HarbergerMarket.sol [hunk 3]
```diff
@@ -11,7 +12,7 @@ import "./AccessRoles.sol";
 /**
  * @dev Market place with Harberger tax. Market attaches one ERC20 contract as currency.
  */
-contract HarbergerMarket is ERC721, IHarbergerMarket, Multicall, AccessRoles {
+contract HarbergerMarket is ERC721, IHarbergerMarket, Multicall, AccessRoles, ERC721Enumerable {
```
Comment (guoliu) [inline src/TheSpace/HarbergerMarket.sol:15]: We can replace ERC721 with ERC721Enumerable?

File: src/TheSpace/HarbergerMarket.sol [hunk 4]
```diff
@@ -103,12 +104,20 @@ contract HarbergerMarket is ERC721, IHarbergerMarket, Multicall, AccessRoles {
         public
         view
         virtual
-        override(AccessControl, ERC721, IERC165)
+        override(AccessControl, ERC721, IERC165, ERC721Enumerable)
```
Comment (guoliu) [inline src/TheSpace/HarbergerMarket.sol:107]: Same here, I think we can replace ERC721 with ERC721Enumerable

File: src/TheSpace/HarbergerMarket.sol [hunk 5]
```diff
@@ -103,12 +104,20 @@ contract HarbergerMarket is ERC721, IHarbergerMarket, Multicall, AccessRoles {
...
         returns (bool)
     {
         return super.supportsInterface(interfaceId_);
     }
 
+    function _beforeTokenTransfer(
```
Comment (guoliu) [inline src/TheSpace/HarbergerMarket.sol:113]: If we replace ERC721 with ERC721Enumerable, we can remove this as well.

**Review by guoliu — CHANGES_REQUESTED** [review]
Nice 🍀 Left some comments.

File: src/TheSpace/HarbergerMarket.sol (hunk 5 above)
Comment (gary02) [reply src/TheSpace/HarbergerMarket.sol]: these changes have done.

**Review by guoliu — APPROVED** [review]
Great job 🌿

@gary02 before merging, can you run a `make snapshot` to keep track of the gas fee?

**Comment by gary02** [comment]
> @gary02 before merging, can you run a `make snapshot` to keep track of the gas fee?


sure, but I encounter trouble when run `make snapshot`:
```
Failed tests:
[FAIL. Reason: Invalid data. Counterexample: calldata=0x55a34fd7000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, args=[0, 0, 0]] testTokenURI(uint8,uint8,uint16) (runs: 0, μ: 0, ~: 0)
[FAIL. Reason: Invalid data. Counterexample: calldata=0x65ba286500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, args=[0, 0]] testForkRecursively(uint8,uint96) (runs: 0, μ: 0, ~: 0)
[FAIL. Reason: Invalid data. Counterexample: calldata=0xd15566e6000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, args=[0, 0, 0]] testSplitRoyalty(uint8,uint8,uint96) (runs: 0, μ: 0, ~: 0)

Encountered a total of 3 failing tests, 38 tests succeeded
Makefile:9: recipe for target 'snapshot' failed
make: *** [snapshot] Error 1

```
any tips to fix this?

**Comment by robertu7** [comment]
> > @gary02 before merging, can you run a `make snapshot` to keep track of the gas fee?
> 
> sure, but I encounter trouble when run `make snapshot`:
> 
> ```
> Failed tests:
> [FAIL. Reason: Invalid data. Counterexample: calldata=0x55a34fd7000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, args=[0, 0, 0]] testTokenURI(uint8,uint8,uint16) (runs: 0, μ: 0, ~: 0)
> [FAIL. Reason: Invalid data. Counterexample: calldata=0x65ba286500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, args=[0, 0]] testForkRecursively(uint8,uint96) (runs: 0, μ: 0, ~: 0)
> [FAIL. Reason: Invalid data. Counterexample: calldata=0xd15566e6000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, args=[0, 0, 0]] testSplitRoyalty(uint8,uint8,uint96) (runs: 0, μ: 0, ~: 0)
> 
> Encountered a total of 3 failing tests, 38 tests succeeded
> Makefile:9: recipe for target 'snapshot' failed
> make: *** [snapshot] Error 1
> ```
> 
> any tips to fix? @guoliu

What's your Foundry version?

```bash
➜  contracts git:(develop) forge
forge 0.2.0 (95857c9 2022-04-11T00:19:01.219885+00:00)
Build, test, fuzz, formally verify, debug & deploy solidity contracts.

USAGE:
    forge <SUBCOMMAND>
```

**Comment by gary02** [comment]
`make snapshot` is ok now after update foundry from 0.1.0 to 0.2.0.   Thx @robertu7

**Comment by gary02** [comment]
I find that `make snapshot` do not include fuzz tests by default. 
 run `forge snapshot --gas-report --include-fuzz-tests` will fail:
```
Failed tests:
[FAIL. Counterexample: calldata=0x55a34fd7000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001, args=[1, 1, 1]] testTokenURI(uint8,uint8,uint16) (runs: 0, μ: 0, ~: 0)

Encountered a total of 1 failing tests, 40 tests succeeded
```

My Foundry version now:
```
$ forge 
 forge 0.2.0 (b3fe652 2022-04-18T07:41:40.363902807+00:00)
```

@robertu7

**Comment by gary02** [comment]
> I find that `make snapshot` do not include fuzz tests by default. run `forge snapshot --gas-report --include-fuzz-tests` will fail:
> 
> ```
> Failed tests:
> [FAIL. Counterexample: calldata=0x55a34fd7000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001, args=[1, 1, 1]] testTokenURI(uint8,uint8,uint16) (runs: 0, μ: 0, ~: 0)
> 
> Encountered a total of 1 failing tests, 40 tests succeeded
> ```
> 
> My Foundry version now:
> 
> ```
> $ forge 
>  forge 0.2.0 (b3fe652 2022-04-18T07:41:40.363902807+00:00)
> ```
> 
> which installed by `cargo install --git https://github.com/foundry-rs/foundry --locked` @robertu7

I figure out that testTokenURI failed  because of  missing 'ehthers' in node env. After run `npm install .`, all tests is ok now.

**Pull request merged**